dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java

```diff
@@ -3474,6 +3486,146 @@ protected void checkLogsForErrorsAndWarnings(List<String> failuresRegExps, Strin
+        if (action != null && action instanceof FeatureReplacementAction) {
+            FeatureReplacementAction featureReplacementAction = (FeatureReplacementAction) action;
+
+            //only check the features if this server was included in the FeatureReplacementAction
+            String serverName = getServerName();
+
+            Set<String> servers = featureReplacementAction.getServers();
+            //only check the features if the FeatureReplacementAction applies to this server
+            if (!servers.isEmpty() &&
+                !servers.contains(FeatureReplacementAction.NO_SERVERS) &&
+                (servers.contains(serverName) || servers.contains(FeatureReplacementAction.ALL_SERVERS))) {
+
+                Set<String> expectedFeatures = new HashSet<>(featureReplacementAction.getAddFeatures()); //the expected features, if present
+                expectedFeatures.addAll(featureReplacementAction.getAlwaysAddFeatures());
+                Set<String> installedFeatures = getInstalledFeatures(); //the features actually installed at runtime
+
+                //compare each installed feature to the expected ones
+                for (String installedFeature : installedFeatures) {
+                    //ignore versionless features
+                    int dash = installedFeature.indexOf("-");
+                    if (dash > -1) {
+                        String versionlessInstalledFeature = removeFeatureVersion(installedFeature);
+                        for (String replacementFeature : expectedFeatures) {
+                            String versionlessReplacementFeature = removeFeatureVersion(replacementFeature);
+                            if (versionlessReplacementFeature.equalsIgnoreCase(versionlessInstalledFeature)) {
+                                //if the base feature name matches but the version does not, throw an exception.
+                                if (!replacementFeature.equalsIgnoreCase(installedFeature)) {
+                                    String message = "Runtime feature was not of the expected version for repeat action (" + serverName + ", " + action.getID() + "). "
+                                                     + "Expected: " + replacementFeature + ", Actual: " + installedFeature + ". "
+                                                     + "This is caused by a feature not being explicitly set in the FAT's server.xml such that FeatureReplacementAction does not replace it properly.";
+
+                                    //check for exempt servers should eventually be removed
+                                    //if exempt then output info message, otherwise throw exception
+                                    if (EXEMPT_SERVERS_SET.contains(serverName)) {
+                                        Log.info(c, method, message);
+                                    } else {
+
+                                        throw new Exception(message);
+                                    }
```

Comment:
This will only report the first discrepency found, is that enough or should we collect all of them?

Reply:
I changed this code to collect all the discrepancies and output a single message at the end.